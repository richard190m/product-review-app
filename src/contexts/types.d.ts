import { ReactNode } from 'react';

// Reviews
export interface ReviewProviderProps {
  children: ReactNode;
}

export interface Review {
  name: string;
  email: string;
  rating: number;
  title: string;
  comment: string;
  date: Date;
}

export interface HandleSubmit {
  // eslint-disable-next-line no-unused-vars
  (review: Review): void;
}

export interface ReviewsContextInterface {
  reviews: Review[];
  submit: HandleSubmit;
}

// Breakpoint
export interface BreakpointProviderProps {
  children: ReactNode;
}

export interface BreakpointContextInterface {
  breakpoint: string;
}
