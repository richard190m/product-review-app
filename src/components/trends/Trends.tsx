import { ChartsContainer, Menu, MenuButton, TrendsTitle } from './styles';
import React, { useState } from 'react';
import AllReviewsChart from './all-reviews-chart';
import PropTypes from 'prop-types';
import SixMonthsChart from './six-months-chart';
import { TrendsProps } from './types';

const Charts = ['allReviews', '6months'];

const Trends = ({ className }: TrendsProps) => {
  const [chartToShow, setChartToShow] = useState(Charts[0]);

  const handleClick = (chart: string) => {
    setChartToShow(chart);
  };

  return (
    <div className={className}>
      <TrendsTitle>Trends</TrendsTitle>
      <Menu>
        <MenuButton
          onClick={() => handleClick(Charts[0])}
          disabled={chartToShow === Charts[0]}
          data-testid="all-reviews-button"
        >
          All reviews
        </MenuButton>
        <MenuButton
          onClick={() => handleClick(Charts[1])}
          disabled={chartToShow === Charts[1]}
          data-testid="6months-button"
        >
          6 Months Chart
        </MenuButton>
      </Menu>
      <ChartsContainer>
        {chartToShow === Charts[0] ? <AllReviewsChart /> : <SixMonthsChart />}
      </ChartsContainer>
    </div>
  );
};

Trends.propTypes = {
  className: PropTypes.string,
};

export default Trends;
