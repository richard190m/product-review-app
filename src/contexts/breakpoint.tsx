import { BreakpointContextInterface, BreakpointProviderProps } from './types';
import { createContext, useEffect, useState } from 'react';

export const BreakpointContext =
  createContext<BreakpointContextInterface | null>(null);

export const BreakpointProvider = ({ children }: BreakpointProviderProps) => {
  const [breakpoint, setBreakpoint] = useState<string>('');

  const handleBreakpoint = () => {
    if (window.innerWidth < 768) breakpoint != 'xs' && setBreakpoint('xs');
    else if (window.innerWidth < 1024)
      breakpoint != 'sm' && setBreakpoint('sm');
    else if (window.innerWidth < 1440)
      breakpoint != 'md' && setBreakpoint('md');
    else breakpoint != 'lg' && setBreakpoint('lg');
  };

  useEffect(() => {
    handleBreakpoint();

    function handleResize() {
      handleBreakpoint();
    }

    window.addEventListener('resize', handleResize);

    return () => window.removeEventListener('resize', handleResize);
  }, []);

  return (
    <BreakpointContext.Provider value={{ breakpoint }}>
      {children}
    </BreakpointContext.Provider>
  );
};
