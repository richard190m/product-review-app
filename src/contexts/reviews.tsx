import { createContext, useEffect, useState } from 'react';
import {
  HandleSubmit,
  Review,
  ReviewProviderProps,
  ReviewsContextInterface,
} from './types';
import MockReviews from '../mocks/MockReviews';

export const ReviewsContext = createContext<ReviewsContextInterface | null>(
  null,
);

export const ReviewsProvider = ({ children }: ReviewProviderProps) => {
  const [reviews, setReviews] = useState<Review[]>([]);

  const handleSubmit: HandleSubmit = review => {
    setReviews([...reviews, review]);
  };

  useEffect(() => {
    // Load Reviews (mocks)
    setReviews(MockReviews);
  }, []);

  return (
    <ReviewsContext.Provider value={{ reviews, submit: handleSubmit }}>
      {children}
    </ReviewsContext.Provider>
  );
};
