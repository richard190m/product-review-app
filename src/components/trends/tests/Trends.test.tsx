import { cleanup, fireEvent, render } from '@testing-library/react';
import { ReviewsProvider } from '../../../contexts/reviews';
import Trends from '../Trends';

const renderTrends = () =>
  render(
    <ReviewsProvider>
      <Trends />
    </ReviewsProvider>,
  );

describe('Trends', () => {
  afterEach(cleanup);
  test('should render correctly', () => {
    const { container, getByTestId } = renderTrends();

    expect(container).toMatchSnapshot();
    expect(getByTestId('doughnut-chart')).toBeInTheDocument();
  });

  test('should show line chart when click on menu 6 months chart', () => {
    const { getByTestId } = renderTrends();

    // Click to show 6 months chart
    fireEvent.click(getByTestId('6months-button'));

    expect(getByTestId('line-chart')).toBeInTheDocument();
  });
});
