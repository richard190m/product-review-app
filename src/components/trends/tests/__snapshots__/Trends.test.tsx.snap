// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Trends should render correctly 1`] = `
<div>
  <div>
    <h2
      class="sc-dkzDqf ia-DSOb"
    >
      Trends
    </h2>
    <div
      class="sc-bczRLJ emMGHK"
    >
      <button
        class="sc-gsnTZi cYybaM"
        data-testid="all-reviews-button"
        disabled=""
      >
        All reviews
      </button>
      <button
        class="sc-gsnTZi cYybaM"
        data-testid="6months-button"
      >
        6 Months Chart
      </button>
    </div>
    <div
      class="sc-hKMtZM kOHBsS"
    >
      <div
        data-testid="doughnut-chart"
      >
        Doughnut Chart
      </div>
    </div>
  </div>
</div>
`;
